pkg/core/extension_store.go

```diff
@@ -0,0 +1,578 @@
+	extension *extensionRecord
+	// Map of endpoints associated with the service, indexed by URL
+	endpoints map[string]*extensionEndpointRecord
+	// Map of credentials associated with the service, indexed by ID
+	credentials map[string]*extensionCredentialsRecord
+}
+
+type extensionEndpointRecord struct {
+	domain.ExtensionEndpoint
+	// parent reference
+	service *extensionServiceRecord
+}
+
+type extensionCredentialsRecord struct {
+	domain.ExtensionCredentials
+	// parent reference
+	service *extensionServiceRecord
+}
+
+// ExtensionStore is an in memory store for extensions
+type ExtensionStore struct {
+	// map of extensions indexed by ID
+	items map[string]*extensionRecord
+}
+
+// NewExtensionStore returns an in-memory extension store instance
+func NewExtensionStore() *ExtensionStore {
+	return &ExtensionStore{
+		items: make(map[string]*extensionRecord),
+	}
+}
+
+// simple unique extension ID generator
+func (store *ExtensionStore) allocateExtensionID(extension *domain.Extension) string {
+	prefix := extension.Product
+	if prefix != "" {
+		prefix = prefix + "-"
+	}
+	for {
+		ID := prefix + randstr.String(8)
+		if store.items[ID] == nil {
+			return ID
+		}
+	}
+}
+
+// simple unique extension service ID generator
+func (store *ExtensionStore) allocateExtensionServiceID(extension *extensionRecord, service *domain.ExtensionService) string {
+	prefix := service.Resource
+	if prefix == "" && extension.Product != "" {
+		prefix = extension.Product + "-service"
+	}
+	if prefix != "" {
+		prefix = prefix + "-"
+	}
+	for {
+		ID := prefix + randstr.String(8)
+		if extension.services[ID] == nil {
+			return ID
+		}
+	}
+}
+
+// simple unique extension credentials ID generator
+func (store *ExtensionStore) allocateExtensionCredentialsID(service *extensionServiceRecord, credentials *domain.ExtensionCredentials) string {
+	prefix := service.Resource
+	if prefix == "" {
+		prefix = "creds"
+	}
+	if prefix != "" {
+		prefix = prefix + "-"
+	}
+	for {
+		ID := prefix + randstr.String(8)
+		if service.credentials[ID] == nil {
+			return ID
+		}
+	}
+}
+
+// StoreExtension - store an extension
+func (store *ExtensionStore) StoreExtension(ctx context.Context, extension *domain.Extension) (result *domain.Extension, err error) {
+	if extension.ExtensionID == "" {
+		extension.ExtensionID = store.allocateExtensionID(extension)
+	}
+
+	if store.items[extension.ExtensionID] != nil {
+		return nil, domain.NewErrExtensionExists(extension.ExtensionID)
+	}
+
+	// store a copy of the input extension
+	store.items[extension.ExtensionID] = &extensionRecord{*extension, make(map[string]*extensionServiceRecord)}
+	return extension, nil
+}
+
+// StoreService - store an extension service
+func (store *ExtensionStore) StoreService(ctx context.Context, service *domain.ExtensionService) (result *domain.ExtensionService, err error) {
+	extRecord := store.items[service.ExtensionID]
+	if extRecord == nil {
+		return nil, domain.NewErrExtensionNotFound(service.ExtensionID)
+	}
```

Comment:
As a rule of thumb, I'm never storing structs as values instead of pointers unless I really have to, because it's inefficient, as it performs a full-struct-copy operation on assignment. So I don't see the point in doing this.

Reply:
Yeah, I thought the same. But turns out that's not always the case and there are many factors to consider to reach a conclusion, see:

- https://philpearl.github.io/post/bad_go_slice_of_pointers/
- https://groups.google.com/g/golang-nuts/c/086NTNXus2E
- https://medium.com/a-journey-with-go/go-should-i-use-a-pointer-instead-of-a-copy-of-my-struct-44b43b104963

TBH, I am not really sure what fits better in this case.

Reply:
As discussed, this makes sense when the allocated heap is released when the entire goroutine exits, but in this case, the heap memory remains allocated as a consequence of the "in-memory" store functionality.